server/storage/wal/version.go

```diff
@@ -59,6 +75,12 @@ func etcdVersionFromData(entryType raftpb.EntryType, data []byte) *semver.Versio
 			return nil
 		}
 		msg = proto.MessageReflect(&raftReq)
+		if raftReq.ClusterVersionSet != nil {
+			ver, err = semver.NewVersion(raftReq.ClusterVersionSet.Ver)
+			if err != nil {
+				panic(err)
+			}
+		}
```

Comment:
@serathius You only check `ClusterVersionSet` here,  I am curious why we don't check other message type?

Reply:
If you look below at the return function we not only check `ClusterVersionSet` we pick max with `etcdVersionFromMessage(msg)`. The `etcdVersionFromMessage` checks etcd_version proto tags on WAL which is meant to be the main mechanism to check version, based on non-empty proto fields. Check for `ClusterVersionSet` is the only case that we also check contents of the proto field.

Unique handling of `ClusterVersionSet` comes from assumption for online downgrade that we shouldn't allow downgrading etcd if there was a period of higher version in WAL that was not covered by snapshot. This is ok for online downgrade as we snapshot as part of the process.

Reply:
thx for the clarification.

> we shouldn't allow downgrading etcd if there was a period of higher version in WAL that was not covered by snapshot.

Please take a look at https://github.com/etcd-io/etcd/pull/19263, thx.

Reply:
My suggestion would be to remove this code and consider adding a separate field to WAL for testing.